packages/eslint-plugin/tests/rules/prefer-promise-reject-errors.test.ts

```diff
@@ -0,0 +1,227 @@
+import { RuleTester } from '@typescript-eslint/rule-tester';
+import type { TSESLint } from '@typescript-eslint/utils';
+import { AST_NODE_TYPES } from '@typescript-eslint/utils';
+
+import type {
+  MessageIds,
+  Options,
+} from '../../src/rules/prefer-promise-reject-errors';
+import rule from '../../src/rules/prefer-promise-reject-errors';
+import { getFixturesRootDir } from '../RuleTester';
+
+const rootDir = getFixturesRootDir();
+const ruleTester = new RuleTester({
+  parserOptions: {
+    ecmaVersion: 2015,
+    tsconfigRootDir: rootDir,
+    project: './tsconfig.json',
+  },
+  parser: '@typescript-eslint/parser',
+});
+
+ruleTester.run('prefer-promise-reject-errors', rule, {
+  valid: [
+    'Promise.resolve(5);',
+    {
+      code: 'Promise.reject();',
+      options: [
+        {
+          allowEmptyReject: true,
+        },
+      ],
+    },
+    'Promise.reject(new Error());',
+    'Promise.reject(new TypeError());',
+    "Promise.reject(new Error('foo'));",
+    `
+      class CustomError extends Error {}
+      Promise.reject(new CustomError());
+    `,
+    `
+      declare const foo: () => { err: SyntaxError };
+      Promise.reject(foo().err);
+    `,
+    `
+      declare const foo: () => Promise<Error>;
+      Promise.reject(await foo());
+    `,
+    'Promise.reject((foo = new Error()));',
+    `
+      const foo = Promise;
+      foo.reject(new Error());
+    `,
+    "Promise['reject'](new Error());",
+
+    `
+      new Promise(function (resolve, reject) {
+        resolve(5);
+      });
+    `,
+    `
+      new Promise(function (resolve, reject) {
+        reject(new Error());
+      });
+    `,
+    `
+      new Promise((resolve, reject) => {
+        reject(new Error());
+      });
+    `,
+    'new Promise((resolve, reject) => reject(new Error()));',
+    {
+      code: `
+        new Promise(function (resolve, reject) {
+          reject();
+        });
+      `,
+      options: [
+        {
+          allowEmptyReject: true,
+        },
+      ],
+    },
+    'new Promise();',
+    'new Promise(5);',
```

Comment:
those 2 are not valid code regardless of the rule

Reply:
I just added them to improve test coverage. This is a valid syntax, although it throws an exception in runtime

Reply:
No need to include these IMO, the rule doesn't really check for them and they'd give a runtime + TypeScript complaint.

Reply:
Ah if we remove these 2 cases, the coverage is no longer 100% :(
<img src="https://github.com/typescript-eslint/typescript-eslint/assets/61150013/e29d16e6-379e-45ed-ac88-cc6dd9154d3f" alt="100% coverage">


This branch won't be tested: https://github.com/typescript-eslint/typescript-eslint/pull/8011/files#diff-801508874bd1d1cb9f122a9ca577fdf2350bfd756c5e63572487b65b859d6799R112-R115

But if we remove this branch, the rule will throw runtime error during linting `new Promise()`. I suppose it shouldn't break on valid syntax
Is `/* istanbul ignore */` applicable for such a case?

Reply:
`/* istanbul ignore */` is almost never appropriate (even most or all of our existing cases should probably be replaced with assertions or general-use utilities). 

If the rule would _crash_ in a particular case then I guess including the test makes sense yeah.